src/openrc-shutdown/broadcast.c

```diff
@@ -193,10 +193,19 @@ void broadcast(char *text)
 			}
 		}
 		alarm(0);
-		if (fd >= 0)
+		/*
+		 * It is possible we got here because fputs or fflush got
+		 * interrupted by SIGALARM. Close the file descriptor first to
```

Comment:
Doesn't this result in a double close of `fd`?
If blocking is the issue why not make the fd non-blocking?

Reply:
Yes, it does result in a double close. I should update the summary: the bug is a double fclose, since that causes a double free.

The fd is already non-blocking, but the original author doesn't trust the kernel to respect that apparently. The whole SIGALARM path should really be stripped out and simplified.